src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1112,23 +1112,40 @@ FinalizeDistributedPlan(DistributedPlan *plan, Query *originalQuery)
 static void
 RecordSubPlansUsedInPlan(DistributedPlan *plan, Query *originalQuery)
 {
-	/* first, get all the subplans in the query */
-	plan->usedSubPlanNodeList = FindSubPlansUsedInNode((Node *) originalQuery);
+	Node *havingQual = originalQuery->havingQual;
+
+	/* temporarily set to NULL, we're going to restore before the function returns */
+	originalQuery->havingQual = NULL;
 
 	/*
-	 * Later, remove the subplans used in the HAVING clause, because they
-	 * are only required in the coordinator. Including them in the
-	 * usedSubPlanNodeList prevents the intermediate results to be sent to the
-	 * coordinator only.
+	 * Mark the subplans as needed on remote side. Note that this decision is revisited
+	 * on execution, when the query only consists of intermediate results.
 	 */
+	List *subplansExceptHaving = FindSubPlansUsedInNode((Node *) originalQuery);
+	SetUsedPlanListLocation(subplansExceptHaving, RequiredOnlyRemote);
+
+	/* do the same for HAVING part of the query */
+	List *subplansInHaving = NIL;
 	if (originalQuery->hasSubLinks &&
-		FindNodeCheck(originalQuery->havingQual, IsNodeSubquery))
+		FindNodeCheck(havingQual, IsNodeSubquery))
 	{
-		List *subplansInHaving = FindSubPlansUsedInNode(originalQuery->havingQual);
-
-		plan->usedSubPlanNodeList =
-			list_difference(plan->usedSubPlanNodeList, subplansInHaving);
+		subplansInHaving = FindSubPlansUsedInNode(havingQual);
+		SetUsedPlanListLocation(subplansInHaving, plan->masterQuery ? RequiredOnlyLocal : RequiredBothRemoteAndLocal);
```

Comment:
See my response as a general reply to `masterQuery` notes you have in the PR.

The point where we check for `masterQuery` is recursively done per subquery.  It means that you could have extremely complex SQL with lots of `HAVING` clauses many arbitrary subqueries, where this logic may or may not not kick in.

This logic is only when the top level query that is being planned has subquery in  `HAVING` clause, such as

```SQL
SELECT key FROM table_1 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_1)
DEBUG:  Router planner cannot handle multi-shard select queries
DEBUG:  Router planner cannot handle multi-shard select queries
DEBUG:  generating subplan 38_1 for subquery SELECT max(value) AS max FROM locally_execute_intermediate_results.table_1
DEBUG:  Plan 38 query after replacing subqueries and CTEs: SELECT key FROM locally_execute_intermediate_results.table_1 GROUP BY key HAVING (max(value) OPERATOR(pg_catalog.>) (SELECT intermediate_result.max FROM read_intermediate_result('38_1'::text, 'binary'::citus_copy_format) intermediate_result(max text)))
DEBUG:  Router planner cannot handle multi-shard select queries
DEBUG:  Subplan 38_1 will be written to local file
┌─────┐
│ key │
├─────┤
└─────┘
(0 rows)
```

However, when you have more complex SQL where any subquery has HAVING subquery, we could get into the same situation such as the below where subPlan `1_2` literally has the same structure where the `master` query has `HAVING` subquery so `1_1` is written to locally:

```SQL
 SELECT * FROM                         
  (SELECT key,random() FROM table_1 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_1) LIMIT 1) as foo;
DEBUG:  generating subplan 1_1 for subquery SELECT max(value) AS max FROM locally_execute_intermediate_results.table_1
DEBUG:  generating subplan 1_2 for subquery SELECT key, random() AS random FROM locally_execute_intermediate_results.table_1 GROUP BY key HAVING (max(value) OPERATOR(pg_catalog.>) (SELECT intermediate_result.max FROM read_intermediate_result('1_1'::text, 'binary'::citus_copy_format) intermediate_result(max text))) LIMIT 1
DEBUG:  Plan 1 query after replacing subqueries and CTEs: SELECT key, random FROM (SELECT intermediate_result.key, intermediate_result.random FROM read_intermediate_result('1_2'::text, 'binary'::citus_copy_format) intermediate_result(key integer, random double precision)) foo
DEBUG:  Subplan 1_1 will be written to local file
DEBUG:  Subplan 1_2 will be written to local file
NOTICE:  executing the command locally: SELECT key, random FROM (SELECT intermediate_result.key, intermediate_result.random FROM read_intermediate_result('1_2'::text, 'binary'::citus_copy_format) intermediate_result(key integer, random double precision)) foo
┌─────┬────────┐
│ key │ random │
├─────┼────────┤
└─────┴────────┘
(0 rows)

Time: 528.945 ms
```

In general, I'm planning to add some more tests around triggering this case, including with JOINs with subqueries/CTEs as well to force for 


> This could also solve my concerns above, but would require a new walker. Instead of finding the subplans we can have a specialized walker that could set the bits on the sub plans, and increases the bitmask with the having bit when it recurses into a having clause. Although this might be a significant change.

Well, such a walker would be pretty hard (and prone to bugs) to write since we rely on `query_tree_walker` and we don't care how it handles the traversals. Doing something special for `HAVING` could be tricky, especially considering all the cases where Postgres allows subqueries in weird  places :) 


Finally, as you noted below, we can not easily pushdown subqueries in HAVING clause, because it fails on the colocation checks, we always need to plan recursively to succeed in planning
```SQL

  SELECT * FROM 
  (SELECT key FROM table_1 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_1 GROUP BY key)) as foo,
  (SELECT key FROM table_2 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_2 GROUP BY key)) as bar
  WHERE foo.key = bar.key;
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
```

But, if you have something like the following, we can push it down (HAVING subqueries has correlation from the above query);
```SQL
  SELECT * FROM 
  (SELECT key FROM table_1 as t1 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_1 WHERE table_1.key = t1.key GROUP BY key)) as foo,
  (SELECT key FROM table_2 as t2 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_2 WHERE table_2.key = t2.key GROUP BY key)) as bar
  WHERE foo.key = bar.key;
```

And, correlated subqueries fail if they are not pushdownable (second HAVING is grouped on `value`):
```SQL
 SELECT * FROM 
  (SELECT key FROM table_1 as t1 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_1 WHERE table_1.key = t1.key GROUP BY key)) as foo,
  (SELECT key FROM table_2 as t2 GROUP BY key HAVING max(value) > (SELECT max(value) FROM table_2 WHERE table_2.key = t2.key GROUP BY value)) as bar
  WHERE foo.key = bar.key;
ERROR:  cannot push down this subquery
DETAIL:  Group by list without partition column is currently unsupported when a subquery references a column from another query
```